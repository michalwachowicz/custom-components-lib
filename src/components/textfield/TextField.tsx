import { useId, useState } from "react";
import * as styles from "./TextField.module.scss";
import type { InputVariant } from "../../types/InputVariant";

interface TextFieldProps {
  label: string;
  variant?: InputVariant;
  disabled?: boolean;
  error?: boolean | string;
}

export default function TextField({
  label,
  variant = "standard",
  disabled = false,
  error = false,
}: TextFieldProps) {
  const id = useId();

  const [focused, setFocused] = useState(false);
  const [value, setValue] = useState("");

  const focusedClass = focused ? styles["textfield-focused"] : "";
  const notEmptyClass = value !== "" ? styles["textfield-not-empty"] : "";
  const errorClass = error ? styles["textfield-error"] : "";

  return (
    <div
      className={`${styles[`textfield-container-${variant}`]}`}
      data-testid='textfield-container'
    >
      <div
        data-testid='textfield'
        className={`${styles.textfield} ${
          styles[`textfield-${variant}`]
        } ${focusedClass} ${notEmptyClass} ${errorClass}`}
      >
        <label className={styles["textfield-label"]} htmlFor={id}>
          {label}
        </label>

        <div className={styles["textfield-input-wrapper"]}>
          <input
            className={styles["textfield-input"]}
            type='text'
            id={id}
            disabled={disabled}
            value={value}
            onChange={(e) => setValue(e.target.value)}
            onFocus={() => setFocused(true)}
            onBlur={() => setFocused(false)}
          />
        </div>

        {variant === "outlined" && (focused || value !== "") && (
          <fieldset
            className={styles["textfield-fieldset"]}
            data-testid='textfield-fieldset'
          >
            <legend className={styles["textfield-fieldset-legend"]}>
              {label}
            </legend>
          </fieldset>
        )}
      </div>

      {typeof error === "string" && error.trim() !== "" && (
        <div className={styles["textfield-error-message"]}>{error}</div>
      )}
    </div>
  );
}
