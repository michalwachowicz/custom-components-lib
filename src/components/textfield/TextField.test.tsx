import { act } from "react";
import { render, screen } from "@testing-library/react";
import TextField from "./TextField";
import type { InputVariant } from "../../types/InputVariant";
import userEvent from "@testing-library/user-event";

describe("<TextField />", () => {
  it("inputs text correctly", async () => {
    render(<TextField label='Name' />);

    const input = screen.getByRole("textbox");
    const user = userEvent.setup();

    expect(input).toHaveValue("");

    await act(async () => {
      await user.type(input, "test");
    });

    expect(screen.getByRole("textbox")).toHaveValue("test");
  });

  it("does not allow input when disabled", async () => {
    render(<TextField label='Name' disabled />);

    const input = screen.getByRole("textbox");
    const user = userEvent.setup();

    expect(input).toHaveValue("");
    expect(input).toBeDisabled();

    await act(async () => {
      await user.type(input, "test");
    });

    expect(screen.getByRole("textbox")).toHaveValue("");
  });

  it("renders label correctly", () => {
    render(<TextField label='Name' />);
    expect(screen.getByLabelText("Name")).toBeInTheDocument();
  });

  it("applies error class if error prop is true", () => {
    render(<TextField label='Name' error />);
    expect(screen.getByTestId("textfield")).toHaveClass("textfield-error");
  });

  it("shows error message if error prop is a string", () => {
    render(<TextField label='Name' error='Incorrect input.' />);

    expect(screen.getByTestId("textfield")).toHaveClass("textfield-error");
    expect(screen.getByText("Incorrect input.")).toBeInTheDocument();
  });

  it("applies focused class when input is focused", async () => {
    render(<TextField label='Name' />);

    const input = screen.getByRole("textbox");
    const user = userEvent.setup();
    const focusedClass = "textfield-focused";

    expect(screen.getByTestId("textfield")).not.toHaveClass(focusedClass);

    await act(async () => {
      await user.click(input);
    });

    expect(screen.getByTestId("textfield")).toHaveClass(focusedClass);
  });

  it("shows legend when variant is outlined and focused or not empty", async () => {
    render(
      <div>
        <TextField label='Name' variant='outlined' />
        <div data-testid='outside' />
      </div>
    );

    const input = screen.getByRole("textbox");
    const user = userEvent.setup();

    expect(screen.queryByTestId("textfield-fieldset")).not.toBeInTheDocument();

    await act(async () => {
      await user.click(input);
    });

    expect(screen.getByTestId("textfield-fieldset")).toBeInTheDocument();

    await act(async () => {
      await user.click(screen.getByTestId("outside"));
    });

    expect(screen.queryByTestId("textfield-fieldset")).not.toBeInTheDocument();

    await act(async () => {
      await user.type(input, "test");
      await user.click(screen.getByTestId("outside"));
    });

    expect(screen.getByTestId("textfield-fieldset")).toBeInTheDocument();
  });

  describe("variants", () => {
    const variants = ["standard", "outlined", "filled"];
    const errorStates = [true, false, "error message"];
    const disabledStates = [true, false];

    const testCases = variants.flatMap((variant) =>
      errorStates.flatMap((error) =>
        disabledStates.map((disabled) => [variant, error, disabled])
      )
    );

    test.each(testCases)(
      "renders correctly with variant=%s, error=%s, disabled=%s",
      (variant, error, disabled) => {
        const { asFragment } = render(
          <TextField
            label='Name'
            variant={variant as InputVariant}
            error={error}
            disabled={!!disabled}
          />
        );

        expect(asFragment()).toMatchSnapshot();
      }
    );
  });
});
