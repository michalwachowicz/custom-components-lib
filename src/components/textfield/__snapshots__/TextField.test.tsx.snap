// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> variants renders correctly with variant=filled, error=error message, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-filled"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-filled   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«ro»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          id="«ro»"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="textfield-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=filled, error=error message, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-filled"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-filled   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rn»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          disabled=""
          id="«rn»"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="textfield-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=filled, error=false, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-filled"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-filled   "
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rm»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          id="«rm»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=filled, error=false, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-filled"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-filled   "
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rl»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          disabled=""
          id="«rl»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=filled, error=true, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-filled"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-filled   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rk»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          id="«rk»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=filled, error=true, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-filled"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-filled   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rj»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          disabled=""
          id="«rj»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=outlined, error=error message, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-outlined"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-outlined   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«ri»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          id="«ri»"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="textfield-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=outlined, error=error message, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-outlined"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-outlined   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rh»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          disabled=""
          id="«rh»"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="textfield-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=outlined, error=false, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-outlined"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-outlined   "
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rg»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          id="«rg»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=outlined, error=false, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-outlined"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-outlined   "
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rf»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          disabled=""
          id="«rf»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=outlined, error=true, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-outlined"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-outlined   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«re»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          id="«re»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=outlined, error=true, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-outlined"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-outlined   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rd»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          disabled=""
          id="«rd»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=standard, error=error message, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-standard"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-standard   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rc»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          id="«rc»"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="textfield-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=standard, error=error message, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-standard"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-standard   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«rb»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          disabled=""
          id="«rb»"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="textfield-error-message"
    >
      error message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=standard, error=false, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-standard"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-standard   "
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«ra»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          id="«ra»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=standard, error=false, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-standard"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-standard   "
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«r9»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          disabled=""
          id="«r9»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=standard, error=true, disabled=false 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-standard"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-standard   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«r8»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          id="«r8»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TextField /> variants renders correctly with variant=standard, error=true, disabled=true 1`] = `
<DocumentFragment>
  <div
    class="textfield-container-standard"
    data-testid="textfield-container"
  >
    <div
      class="textfield textfield-standard   textfield-error"
      data-testid="textfield"
    >
      <label
        class="textfield-label"
        for="«r7»"
      >
        Name
      </label>
      <div
        class="textfield-input-wrapper"
      >
        <input
          class="textfield-input"
          disabled=""
          id="«r7»"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
